nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -975,267 +991,174 @@ ngx_http_modsecurity_init_process(ngx_cycle_t *cycle)
     }
 
-    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, "modSecurity: body filter");
-
-    for (cl = in; cl; cl = cl->next) {
-        apr_bucket  *e;
-        ngx_buf_t   *buf = cl->buf;
-        apr_bucket_brigade  *bb = ctx->brigade;
-        off_t size = ngx_buf_size(buf);
-        if (size) {
-            char *data = apr_pmemdup(bb->p, buf->pos, size);
-            if (data == NULL) {
-                return ngx_http_filter_finalize_request(r, 
-                         &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
-            }
-            e = apr_bucket_pool_create(data , size, bb->p, bb->bucket_alloc);
-            if (e == NULL) {
-                return ngx_http_filter_finalize_request(r, 
-                         &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
-            }
-            APR_BRIGADE_INSERT_TAIL(bb, e);
-        }
+    // Read body if not yet read
+    if (!r->request_body) {
+        rc = ngx_http_read_client_request_body(r, ngx_http_modsecurity_body_handler);
 
-        if (buf->last_buf) {
-            last_buf = 1;
-            buf->last_buf = 0;
-            e = apr_bucket_eos_create(bb->bucket_alloc);
-            if (e == NULL) {
-                return ngx_http_filter_finalize_request(r, 
-                         &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
-            }
-            APR_BRIGADE_INSERT_TAIL(bb, e);
-            break;
+        if (rc >= NGX_HTTP_SPECIAL_RESPONSE) {
+            return rc;
         }
 
-        buf->pos = buf->last;
-    }
-
-    if (!last_buf) {
-        return NGX_AGAIN;
+        return NGX_DONE;
     }
 
-    /* last buf has been saved */
-    ctx->complete = 1;
-    modsecSetResponseBrigade(ctx->req, ctx->brigade);
-
-    if (ngx_http_modsecurity_load_headers_in(r) != NGX_OK
-            || ngx_http_modsecurity_load_headers_out(r) != NGX_OK) {
-
-        return ngx_http_filter_finalize_request(r, 
-                 &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
-    }
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, "modSecurity: handler");
 
-    rc = ngx_http_modsecurity_status(r, modsecProcessResponse(ctx->req));
+    // Get module request context or create if not yet created
+    ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
+    if (ctx == NULL) {
+        ctx = ngx_http_modsecurity_create_ctx(r);
+        if (ctx == NULL) {
+            return NGX_HTTP_INTERNAL_SERVER_ERROR;
+        }
 
-    if (rc != NGX_DECLINED) {
-        return ngx_http_filter_finalize_request(r, &ngx_http_modsecurity, rc);
+        ngx_http_set_ctx(r, ctx, ngx_http_modsecurity);
     }
 
-    apr_brigade_length(ctx->brigade, 0, &content_length);
-
-    rc = move_brigade_to_chain(ctx->brigade, &out, r->pool);
-    if (rc == NGX_ERROR) {
-        return ngx_http_filter_finalize_request(r, 
-                 &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
+    if (ctx->thread_running) {
```

Comment:
why would thread_running set to 1 here, this is the entry point?

Reply:
Adding this comment
// Sometimes Nginx calls ngx_http_modsecurity_handler multiple times for the same request, after a worker thread has already been started. This is to guard against it.